HW-1 : 영화 검색 프로젝트

camai opened the pull request:
Pull Request 수정하여 다시 요청 합니다. 

네이버 검색 api를 이용한 영화 검색 프로젝트 수정
Retrofit을 이용했습니다.
저도 누락된 스펙 있으면 알려주시면 감사하겠습니다!

Cmaffin committed:
Project Setup (gradle-dependencies, AndroidManifest-permission)

Cmaffin committed:
Movie Data Model 추가

Cmaffin committed:
retorofit , okhttp logging, network 추가

Cmaffin committed:
main. xml 수정, movie_item.xml 추가, string.xml 수정

Cmaffin committed:
Movie RecyclerView Adapter 추가

Cmaffin committed:
Main Activity 검색 결과 리사이클러뷰에 보이도록 수정

Cmaffin committed:
kotlin 코드 컨벤션 적용으로 전체 클래스 수정

Sangmeebee commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/adapter/MovieSearchAdapter.kt:
```diff
@@ -0,0 +1,81 @@
+
+    override fun getItemCount(): Int {
+        return movieInfoArrayList.count()
+    }
```
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/799/commits/b77d6a054518cfa81871f7b11026744c8eb1b956
override fun getItemCount() = movieInfoArrayList.count() 
이런식으로 하면 좀 더 깔끔 할 것 같습니다^^

camai replied:
네~ 리뷰해주신것 처럼 진행하면 더 깔끔할 것 같아 수정하여 반영하겠습니다.

Sangmeebee commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/adapter/MovieSearchAdapter.kt:
```diff
@@ -0,0 +1,81 @@
+    }
+    //  Update Movie List Data
+    fun setClearAndAddList( movielist : ArrayList<Items>){
+        with(this.movieInfoArrayList){
```
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/799/commits/b77d6a054518cfa81871f7b11026744c8eb1b956
this 쓰지 않아도 괜찮지 않을까요?

camai replied:
네~ 확인해보니 this를 쓰지 않아도 되네요~ 리뷰 감사합니다. 수정하여 반영하겠습니다.

Sangmeebee commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/MainActivity.kt:
```diff
@@ -1,11 +1,125 @@
 
 class MainActivity : AppCompatActivity() {
+    //  Movie Search Service Create Init
+    private val api = MovieApiService.create()
```
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/799/commits/8018df4f2518a9fbb88d09357b86f6b91da4683b
getMoiveSearchCall() 메소드에서만 사용함으로 메소드 안에서 MovieApiService.create()를 사용하여 RetrofitBuilder를 생성해 주는게 좋을 것 같습니다!

camai replied:
리뷰하시 내용과 같이 굳이 다른 곳에서 val api 를 통한  MovieApiService.create()을 실행하지 않기에 말씀해주신 내용으로 바꾸는게 좀 더 간결하게 사용가능하네요
감사합니다. 수정하여 반영하겠습니다.

Sangmeebee replied:
고생하셨습니다

Sangmeebee reviewed (commented):
저보다 설계를 잘하시고 짠 코드라 오히려 리뷰하면서 도움이 되었습니다.
제꺼에서 수정해야할 것들이 보이네요 ㅎㅎ... 
몇가지 이러면 좋겠다 라는 것만 남겨요~!

Cmaffin committed:
b77d6a0 Commit 에 Code Review 내용에 대한 수정 내역 반영
 - getItemCount() 코드 간결화
 - with(movieInfoArrayList)  this. 코드 삭제

Cmaffin committed:
8018df4 Commit에 대한 Code Review 내용에 대한 수정 내역 반영
- 수정 내역
  : getMoiveSearchCall () 함수에서 MovieApiService.create()를 사용하여 RetrofitBuilder를 생성
  : 사용되지 않은 size 변수 삭제

Cmaffin committed:
UI 수정 및 error img 추가
- 영화 검색 결과 보기에 불편함이 있어 RecyclerView Item.xml 수정 (ImageView 크기 및 scale 수정)
- 영화 검색 결과 보기에 불편함이 있어 출시연도, 감독, 출연자 설명을 위한 string.xml에 내용 추가
- Image Load 시 error 발생 할 경우 대체 이미지 추가

Cmaffin committed:
영화 검색 결과 데이터 모델에서 출시연도 Key 값 오류로 인한 데이터 null 이 발생되어  Key 값 수정 (pubdate -> pubDate)

Sangmeebee reviewed (approved):
고생하셨습니다 승인했습니다
석준님께 리뷰 받으시면 됩니다!

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/MainActivity.kt:
```diff
@@ -1,11 +1,123 @@
+        //  Search Button Click Event
+        btn_search.setOnClickListener(View.OnClickListener {
+            hideKeyboard(this)
+            progressbar.visibility = View.VISIBLE
```
`progressbar.isVisible = true` 하시면 됩니다. 
아래는 반대로 false를 주면 됩니다.

camai replied:
리뷰해주신 isVisible 추적해보니, 제가 사용하고자 한 VISIBLE, GONE 이 boolean으로 사용됨을 확인하였습니다.
수정하여 업로드 하겠습니다.
감사합니다.

camai replied:
34a20c7 에 반영하였습니다.

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/MainActivity.kt:
```diff
@@ -1,11 +1,123 @@
+    //  RecyclerView Adapter Set
+    private fun setAdapterAndRecyclerViewInit() {
+        movieSearchAdapter = MovieSearchAdapter(this)
+        recycler_view.apply {
```
apply 보다 run을 사용해주세요
apply와 run을 언제 사용하면 좋은지 찾아보세요

camai replied:
두개의 차이점을 확인 해보니 사용 용도가 다르네요.
지금 제가 하는 작업은 Recyclerview 개체를 초기화 설정 실행 후 코드를 더이상 사용하지 않기도 하며, 
또한 처리 결과를 리턴 받은 값을 어딘가에 사용하기 위한 작업이 아니기에
리뷰해주신것 같이 run을 사용하여 진행하는게 맞는것으로 확인하였습니다.

수정하여 업로드 하겠습니다. 감사합니다.

camai replied:
b1869bd 에 반영하였습니다.

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/MainActivity.kt:
```diff
@@ -1,11 +1,123 @@
+        movieSearchAdapter = MovieSearchAdapter(this)
+        recycler_view.apply {
+            adapter = movieSearchAdapter
+            layoutManager = LinearLayoutManager(context)
```
layoutManager는 xml에서 추가 할 수 있습니다.

camai replied:
네 확인하여 수정하여 업로드 하겠습니다. 감사합니다.

camai replied:
a2e9f4c 에 반영하였습니다.

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/MainActivity.kt:
```diff
@@ -1,11 +1,123 @@
+        val moviename: String = edit_name.text.toString()
+
+        //  Movie Name Value Check
+        if (moviename.isBlank() && moviename.isNullOrEmpty()) {
```
moviename.isNullOrBlank()면 충분합니다.

camai replied:
네 리뷰 감사합니다.
수정하여 업로드 하겠습니다.

camai replied:
15c7c52 에 반영하였습니다.

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/MainActivity.kt:
```diff
@@ -1,11 +1,123 @@
+
+                    val body = response.body()
+                    body?.let {
+                        runOnUiThread {
```
불필요한 함수입니다.

camai replied:
리뷰를 통해 불필요한 함수를 확인해보니,
Retrofit 을 통한 call.enqueue 비동기로 동작하기 때문에 굳이 runOnUiThread 는 불필요한 함수인게 맞네요.
Ui를 업데이트 한다고 생각만 하여  runOnUiThread를 사용하였네요. 수정하여 업로드 하겠습니다.
감사합니다.

camai replied:
34c5360 에 반영하였습니다.

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/adapter/MovieSearchAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+import com.camai.archtecherstudy.model.Items
+
+class MovieSearchAdapter(
+    private val context: Context
```
context는 생성자로 전달 받을 필요가 없습니다.

camai replied:
0363667 에 반영하였습니다.

context를 생성자로 전달받아 Glide에 넘겨주기 위해 사용하였는데,
확인해보니 itemView에 context가 이미 있다는걸 알게 되어 리뷰주신것 처럼 context 생성자로 전달 받는 부분을 제외하였습니다.

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/adapter/MovieSearchAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+
+    class ViewHolder(itemView: View?) : RecyclerView.ViewHolder(itemView!!) {
+
+        val poster = itemView?.findViewById<ImageView>(R.id.img_poster)
+        val title = itemView?.findViewById<TextView>(R.id.txt_title)
+        val user_rating = itemView?.findViewById<RatingBar>(R.id.rb_user_rating)
+        val pub_data = itemView?.findViewById<TextView>(R.id.txt_pub_data)
+        val director = itemView?.findViewById<TextView>(R.id.txt_director)
+        val actor = itemView?.findViewById<TextView>(R.id.txt_actor)
```
private을 추가해도 될 것 같습니다.

camai replied:
네 해당 변수들은 ViewHolder에서만 사용되기 때문에 private를 추가해주는게 좋을것 같습니다.
수정하여 업로드하겠습니다.

camai replied:
4148d8d 에 반영하였습니다.

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/adapter/MovieSearchAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+            }
+
+            //  Movie Content WebView Call
+            itemView.setOnClickListener {
+                val webpage = Uri.parse(itemInfo.link)
+                val webIntent = Intent(Intent.ACTION_VIEW, webpage)
+                itemView.context.startActivity(webIntent)
+            }
```
click 처리는 onCreateViewHolder에서 하면 됩니다.
onBindViewHolder에서 하면 불필요한 set을 하게 됩니다.

camai replied:
0247909 에 반영하였습니다.

해당 리뷰를 통해 코드를 고치기 위해 많은걸 공부해 볼 수 있었습니다. 감사합니다. 
계속 보면서 좀 더 좋은 방향이 있는지도 검토해보겠습니다.

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/network/MovieApiService.kt:
```diff
@@ -0,0 +1,63 @@
+
+        fun create(): MovieApiService {
+            val httpLoggingInterceptor = HttpLoggingInterceptor()
+            httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
```
BuildConfig.DEBUG가 true일때만 BODY를 넣어주도록 해주세요

camai replied:
수정하여 업로드 하겠습니다.
감사합니다.

camai replied:
c61b931 에 반영하였습니다.

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,62 @@
+
+    </LinearLayout>
+
+    <RelativeLayout
```
RelativeLayout말고 FrameLayout으로 변경해주세요

camai replied:
a3c04d5 에 반영하였습니다.

리뷰 주신 내용의 의미가
RecyclerView 위에 Progress로 로딩을 표현하는 되는데 FrameLayout으로 쉽게 표현 가능하기에 
변경하도록 리뷰 주신게 맞는지요?
실제 사용하는 Layout 중 FrameLayout을 많이 사용하지 않아 리뷰 주신 의도에 대해 확인 시 위와 같은 결론을 얻었습니다.

sjjeong replied:
단순 ui를 겹치기 위한 용도라면 Frame을 쓰셔도 됩니다.

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,62 @@
+    <RelativeLayout
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:layout_marginLeft="16dp"
+        android:layout_marginRight="16dp"
```
left right보단 start end를 사용하도록 해주세요

camai replied:
a3c04d5 에 같이 적용되어 올리게 되었네요.

확인 해보니 레이아웃에 marin으로 적용할때 start end를 사용하고 있네요. 
리뷰 감사합니다.

Cmaffin committed:
HW-1 progress.isVisible true/false 반영

Cmaffin committed:
HW-1 recyclerview init scope apply  -> run  반영

Cmaffin committed:
HW-1 RecyclerView layoutManager xml에서 적용

Cmaffin committed:
HW-1 moviename isNullOrBlank 적용

Cmaffin committed:
HW-1 불필요한 함수 runOnUiThread 제거

Cmaffin committed:
HW-1 ViewHolder view 개체 변수 private 선언 추가

Cmaffin committed:
HW-1 httpLoggingInterceptor Debugging 모드일 때 BODY 추가

Cmaffin committed:
HW-1 RecyclerView Adapter context 생성자 제거

Cmaffin committed:
HW-1 RelativeLayout -> FrameLayout 변경 및 검색 결과 내용 value 표기 변경

Cmaffin committed:
HW-1 onclick event in onCreateViewHolder

sjjeong commented on class13/camai/ArchitectureStudy/app/src/main/java/com/camai/archtecherstudy/network/MovieApiService.kt:
```diff
@@ -0,0 +1,66 @@
+
+            }
+
+            if (BuildConfig.DEBUG)
+                httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
```
라인이 변경되면 {}을 추가해줘야합니다.
간단한것이기 때문에 다음 과제할때 같이 수정해주세요

camai replied:
네 알겠습니다.

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
